### PR TITLE
Simplify ShapeDescriptorIndex

### PR DESCRIPTION
**Pull request by MikeAlhayek**
@sebastienros I think this can slightly help since you are working on #15620

This PR provide a slightly better improvement in the time it takes to create these object and the allocated memory. Here is a benchmark.

image

### CODE REVIEW

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:50]: Can perhaps all of these `SelectMany`s optimized away too? So, have a single iteration instead of 8.

**Comment by sebastienros** [comment]
I think this change doesn't matter at all.

**Comment by MikeAlhayek** [comment]
> I think this change doesn't matter at all.

@sebastienros are you talking about `ToList()` to `ToArray()` original change? It would just save very minor memory for each tenant.

But in the latest commit, reduces the iteration during each tenant's first request.

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs]: @Piedone done.

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:51]: `AddRange() [does this check](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L253) too.

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: Hmm, I'm a bit hesitant due to all the `AddRange`s. Can you do a benchmark? My gut feeling is that it should be better, but still.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: image

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: it seems that a single loop is not ideal. I am guessing the cost of AddRange and expanding the list out weight the cost of iterations.

Comment (Piedone) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: Is `MultipleLoops` the original code? Because then this is a miss, unfortunately. Interesting (and seems a bit fishy) that allocations are exactly the same for all three.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: @Piedone I fixed the benchmark.

image

Comment (Piedone) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: Jus to confirm:

- `SingleLoopLists` is the latest code.
- `MultipelLoopsUsingLists` is the original code.
- `MultipleLoopsArrays` is the code in its first modified form where you only changed Lists to arrays.

Correct? Because then this looks much more realistic.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: @Piedone 
Yes. That is correct definition. I removed the benchmark and cleaned it up. 

But yes, not bad saving for such a minor update with no complexity.

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:54]: Why does this need to be duplicated?

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:64]: Why is this better at the bottom?

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: The benchmark code look useful as well, please re-add it.

Comment (Piedone) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: Great then!

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: @sebastienros requested to remove it :)

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:64]: Order does not matter.

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs]: Good catch. By mistake

File: src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: I highly doubt anyone will run this benchmark again. Data is known now. It's both I've that ensure we don't regress some related component but one that compares two implementations. What I know though is that each and every build of OC would now have to build it too. If people really care they will look at the history of the code, find this PR and do another benchmark.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: @sebastienros I think what @Piedone is saying that way this benchmark was written can be a good example to keep in the benchmark as a code reference. None of the benchmarks show how we can setup a shell and run benchmarks against that shell. The value is in keeping the code sample not the benchmark result. Is that what you are suggesting here Zoltan?

Comment (Piedone) [reply src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs:1]: Yes, but as mentioned, apart from a sample, it'll also be useful for future benchmarking as-is.

**Comment by Piedone** [comment]
Why did you merge this, Mike? It wasn't approved yet.

**Comment by Piedone** [comment]
That's great, but since I've been reviewing this PR, I'd have appreciated not merging before my approval. Not for bureaucracy, but for reviewing code changes since then.

**Comment by MikeAlhayek** [comment]
I understand. Sorry I thought you were good with it since the last thing you had was to bring back that BM. Let me know if you need changes, I can submit a follow up PR

**Pull request merged**
